### PR TITLE
New package: PaRSEC

### PR DESCRIPTION
**Pull request by abouteiller**
PaRSEC: the Parallel Runtime Scheduler and Execution Controller for micro-tasks on distributed heterogeneous systems.

https://icl.utk.edu/dte/

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

### CODE REVIEW

File: var/spack/repos/builtin/packages/parsec/package.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+#
+# Copyright (c) 2017-2020  The University of Tennessee and the University
+#                          of Tennessee Research Foundation.  All rights
+#                          reserved.
+#
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/parsec/package.py]: Why the additional copyright? I think this is unique in the built-in repository. Can we just use the standard Spack header?

File: var/spack/repos/builtin/packages/parsec/package.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    version('master', branch='master')
+    version('3.0.2012-rc1', sha256='a0f013bd5a2c44c61d3d76bab102e3ca3bab68ef2e89d7b5f544b9c1a6fde475')
+    version('1.1.0', sha256='d2928033c121000ae0a554f1e7f757c1f22274a8b74457ecd52744ae1f70b95a', url='https://bitbucket.org/icldistcomp/parsec/get/v1.1.0.tar.bz2')
+
+    variant('build_type', default='RelWithDebInfo', description='CMake build type', values=('Debug', 'Release', ' RelWithDebInfo'))
+    variant('shared', default=True, description='Build a shared library')
+    # We always need MPI for now.
+    # variant('mpi', default=True,
+    #         description='Use MPI for dependency transport between nodes')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/parsec/package.py]: Can you please remove commented variants?

File: var/spack/repos/builtin/packages/parsec/package.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    variant('cuda', default=True, description='Use CUDA for GPU acceleration')
+    variant('profile', default=False, description='Generate profiling data')
+    variant('debug_verbose', default=False, description='Debug version with verbose and paranoid (incurs performance overhead!)')
+    conflicts('+debug_verbose build_type=Release', msg='You need to set build_type=Debug for +debug_verbose')
+    conflicts('+debug_verbose build_type=RelWithDebInfo', msg='You need to set build_type=Debug for +debug_verbose')
+    # SPack does not handle cross-compilation atm
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/parsec/package.py]:
```suggestion
    # TODO: Spack does not handle cross-compilation atm
```

File: var/spack/repos/builtin/packages/parsec/package.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    depends_on('python', when='+profile')
+    depends_on('py-cython', when='+profile')
+    depends_on('py-pandas', when='+profile')
+    depends_on('py-matplotlib', when='+profile')
+    depends_on('py-tables', when='+profile')
+
+    def cmake_args(self):
+        args = [
+            self.define_from_variant('BUILD_SHARED_LIBS', 'shared'),
+            # self.define_from_variant('PARSEC_DIST_WITH_MPI', 'mpi'),
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/parsec/package.py]: I would suggest either having a variant for `mpi` or removing commented code.

**Review by alalazo — CHANGES_REQUESTED** [review]
A few minor issues and a question on additional copyright

**Comment by abouteiller** [comment]
> A few minor issues and a question on additional copyright
All recommended changes have been implemented @alalazo

File: var/spack/repos/builtin/packages/parsec/package.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    list_url    = "https://bitbucket.org/icldistcomp/parsec/downloads/?tab=tags"
+    maintainers = ['abouteiller', 'bosilca', 'herault']
+
+    version('master', branch='master')
+    version('3.0.2012-rc1', sha256='a0f013bd5a2c44c61d3d76bab102e3ca3bab68ef2e89d7b5f544b9c1a6fde475')
+    version('1.1.0', sha256='d2928033c121000ae0a554f1e7f757c1f22274a8b74457ecd52744ae1f70b95a', url='https://bitbucket.org/icldistcomp/parsec/get/v1.1.0.tar.bz2')
+
+    variant('build_type', default='RelWithDebInfo', description='CMake build type', values=('Debug', 'Release', ' RelWithDebInfo'))
+    variant('shared', default=True, description='Build a shared library')
+    variant('cuda', default=True, description='Use CUDA for GPU acceleration')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/parsec/package.py]: Did you consider inheriting also from `CudaPackage`:
```python
class Parsec(CMakePackage, CudaPackage):
```
? That would add automatically:
1. A few variants, like `cuda`
2. Constraints against compilers that are reported in the CUDA documentation

Comment (abouteiller) [reply var/spack/repos/builtin/packages/parsec/package.py]: I am looking into it, how can I set `+cuda` to be the default?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/parsec/package.py]: Just add:
```python
variant('cuda', default=True, description='Build with CUDA')
```
to override the variant default.

File: var/spack/repos/builtin/packages/parsec/package.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    git         = "https://bitbucket.org/icldistcomp/parsec.git"
+    url         = "https://bitbucket.org/icldistcomp/parsec/get/parsec-3.0.2012.tar.bz2"
+    list_url    = "https://bitbucket.org/icldistcomp/parsec/downloads/?tab=tags"
+    maintainers = ['abouteiller', 'bosilca', 'herault']
+
+    version('master', branch='master')
+    version('3.0.2012-rc1', sha256='a0f013bd5a2c44c61d3d76bab102e3ca3bab68ef2e89d7b5f544b9c1a6fde475')
+    version('1.1.0', sha256='d2928033c121000ae0a554f1e7f757c1f22274a8b74457ecd52744ae1f70b95a', url='https://bitbucket.org/icldistcomp/parsec/get/v1.1.0.tar.bz2')
+
+    variant('build_type', default='RelWithDebInfo', description='CMake build type', values=('Debug', 'Release', ' RelWithDebInfo'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/parsec/package.py]:
```suggestion
    variant('build_type', default='RelWithDebInfo', description='CMake build type', values=('Debug', 'Release', 'RelWithDebInfo'))
```

File: var/spack/repos/builtin/packages/parsec/package.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    depends_on('cmake@3.16.0:', type='build')
+    depends_on('python', type='build')
+    depends_on('hwloc')
+    depends_on('mpi')
+    depends_on('papi', when='+profile')
+    depends_on('python', when='+profile')
+    depends_on('py-cython', when='+profile')
+    depends_on('py-pandas', when='+profile')
+    depends_on('py-matplotlib', when='+profile')
+    depends_on('py-tables', when='+profile')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/parsec/package.py]: What deptype are these? Are they only needed at build time, or build/run time?

Comment (abouteiller) [reply var/spack/repos/builtin/packages/parsec/package.py]: build and run when +profile is selected

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/parsec/package.py]: Thanks, just add `type=('build', 'run')` then since the default is build/link.

**Review by adamjstewart — APPROVED** [review]

**Comment by abouteiller** [comment]
All requested changes have been made and passes all tests @alalazo

**Review by alalazo — APPROVED** [review]
Thanks!

**Pull request merged**